Microphones via AOP (+ a few SEP bits).

WhatAmISupposedToPutHere opened the pull request:
This adds the AOP driver, microphone support, lid angle sensor driver, and ambient light sensor driver (disabled pending some m1n1 work). Also contains just enough of a SEP driver to unlock SECDIS on base m1.

Obsoletes #151

Draft, since the mic output is basically useless without correct userspace configs (coming soontm).

jannau committed:
WIP: HID: transport: spi: add Apple SPI transport

Keyboard and trackpad of Apple Sillicon SoCs (M1, M1 Pro/Max) laptops
are are HID devices connected via SPI.

This is the same protocol as implemented by applespi.c. It was not
noticed that protocol is a transport for HID. Adding support for ACPI
based Intel MacBooks will be done in a separate commit.

How HID is mapped in this protocol is not yet fully understood.

Microsoft has a specification for HID over SPI [1] incompatible with the
transport protocol used by Apple.

[1] https://docs.microsoft.com/en-us/windows-hardware/drivers/hid/hid-over-spi

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
HID: magicmouse: use struct input_mt_pos for X/Y

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
HID: magicmouse: use ops function pointers for input functionality

Will be used for supporting MacBook trackpads connected via SPI.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
HID: magicmouse: add support for Macbook trackpads

The trackpads in Macbooks beginning in 2015 are HID devices connected
over SPI. On Intel Macbooks they are currently supported by applespi.c.
This chang adds support for the trackpads on Apple Silicon Macbooks
starting in late 2020. They use a new HID over SPI transport driver.
The touch report format differs from USB/BT Magic Trackpads. It is the
same format as the type 4 format supported by bcm5974.c.

Signed-off-by: Janne Grunau <j@jannau.net>

Commit:
rust: platform: add `ioremap_resource` and `get_resource` methods

This patch adds a logic similar to `devm_platform_ioremap_resource`
function adding:
  - `IoResource` enumerated type that groups the `IORESOURCE_*` macros.
  - `get_resource()` method that is a binding of `platform_get_resource`
  - `ioremap_resource` that is newly written method similar to
    `devm_platform_ioremap_resource`.

Lina: Removed `bit` dependency and rebased

Co-developed-by: Asahi Lina <lina@asahilina.net>
Signed-off-by: Maciej Falkowski <m.falkowski@samsung.com>

marcan committed:
soc: apple: Add RTKit helper driver

This driver can be used for coprocessors that do some background task or
communicate out-of-band, and do not do any mailbox I/O beyond the
standard RTKit initialization.

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
HID: add HOST vendor/device IDs for Apple MTP devices

Apple M2 chips have an embedded MTP processor that handles all HID
functions, and does not go over a traditional bus like SPI. The devices
still have real IDs, so add them here.

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
HID: core: Handle HOST bus type when announcing devices

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
soc: apple: Add DockChannel driver

DockChannel is a simple FIFO interface used to communicate between SoC
blocks. Add a driver that represents the shared interrupt controller for
the DockChannel block, and then exposes probe and data transfer
functions that child device drivers can use to instantiate individual
FIFOs.

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
hid: Add Apple DockChannel HID transport driver

Apple M2 devices have an MTP coprocessor embedded in the SoC that
handles HID for the integrated touchpad/keyboard, and communicates
over the DockChannel interface. This driver implements this new
interface.

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
hid: apple: Bind to HOST devices for MTP

We use BUS_HOST for MTP HID subdevices

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
hid: magicmouse: Add MTP multi-touch device support

Apple M2 devices expose the multi-touch device over the HID over
DockChannel transport, which we represent as the HOST bus type. The
report format is the same, except the legacy mouse header is gone and
there is no enable request needed.

Signed-off-by: Hector Martin <marcan@marcan.st>

hoshinolina committed:
iommu/io-pgtable: Add Apple UAT variant format

Apple Silicon SoCs (M1, M2, etc.) have a GPU with an ARM64 firmware
coprocessor. The firmware and the GPU share page tables in the standard
ARM64 format (the firmware literally sets the base as its TTBR0/1
registers). TTBR0 covers the low half of the address space and is
intended to be per-GPU-VM (GPU user mappings and kernel-managed
buffers), while TTBR1 covers the upper half and is global (firmware
code, data, management structures shared with the AP, and a few
GPU-accessible data structures).

In typical Apple fashion, the permissions are interpreted differently
from traditional ARM PTEs. By default, firmware mappings use Apple SPRR
permission remapping. The firmware only uses that for its own
code/data/MMIO mappings, and those pages are not accessible by the GPU
hardware. We never need to touch/manage these mappings, so this patch
does not support them.

When a specific bit is set in the PTEs, permissions switch to a
different scheme which supports various combinations of firmware/GPU
access. This is the mode intended to be used by AP GPU drivers, and what
we implement here.

The prot bits are interpreted as follows:

- IOMMU_READ and IOMMU_WRITE have the usual meaning.

- IOMMU_PRIV creates firmware-only mappings (no GPU access)
- IOMMU_NOEXEC creates GPU-only structures (no FW access)
- Otherwise structures are accessible by both GPU and FW

- IOMMU_MMIO creates Device mappings for firmware
- IOMMU_CACHE creates Normal-NC mappings for firmware (cache-coherent
  from the point of view of the AP, but slower)
- Otherwise creates Normal mappings for firmware (this requires manual
  cache management on the firmware side, as it is not coherent with the
  SoC fabric)

GPU-only mappings (textures/etc) are expected to use IOMMU_CACHE and are
seemingly coherent with the CPU (or otherwise the firmware/GPU already
issue the required cache management operations when correctly
configured).

There is a GPU-RO/FW-RW mode, but it is not currently implemented (it
doesn't seem to be very useful for the driver). There seems to be no
real noexec control (i.e. for shaders) on the GPU side. All of these
mappings are implicitly noexec for the firmware.

Drivers are expected to fully manage per-user (TTBR0) page tables, but
ownership of shared kernel (TTBR1) page tables is shared between the
firmware and the AP OS. We handle this by simply using a smaller IAS to
drop down one level of page tables, so the driver can install a PTE in
the top-level (firmware-initialized) page table directly and just add an
offset to the VAs passed into the io_pgtable code. This avoids having to
have any special handling for this here. The firmware-relevant data
structures are small, so we do not expect to ever require more VA space
than one top-level PTE covers (IAS=36 for the next level, 64 GiB).

Only 16K page mode is supported. The coprocessor MMU supports huge pages
as usual for ARM64, but the GPU MMU does not, so we do not enable them.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: kernel: platform: Add Device.set_dma_masks()

Allows drivers to configure the DMA masks for a device. Implemented
here, not in device, because it requires a mutable platform device
reference this way (device::Device is a safely clonable reference).

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/gem-shmem: Export VM ops functions

There doesn't seem to be a way for the Rust bindings to get a
compile-time constant reference to drm_gem_shmem_vm_ops, so we need to
duplicate that structure in Rust... this isn't nice...

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: bindings: Add resource_size wrapper

TODO: This isn't abstracted properly yet

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: Add ioremap_np support to io_mem & friends

Apple SoCs require non-posted mappings for MMIO, and this is
automatically handled by devm_ioremap_resource() and friends via a
resource flag. Implement the same logic in kernel::io_mem, so it can
work the same way.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: module_param: Tolerate a trailing newline when parsing

This is the same behavior as kstrtol/kstrtoul, and allows simple
`echo 0 > /sys/module/foo/parameters/bar` commands to work.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: of: Add OF node abstraction

This abstraction enables Rust drivers to walk Device Tree nodes and
query their properties.

Signed-off-by: Asahi Lina <lina@asahilina.net>

jannau committed:
HID: transport: spi: Check status message after transmits

Probably pointless but might be helpful to debug issues.
Gets rid of 'spi_hid_apple_status_ok' is unused compiler warning.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
HID: magicmouse: Add .reset_resume for SPI trackpads

The trackpad has to request multi touch reports during resume.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
HID: transport: spi: Add suspend support

Working suspend and resume. The keyboard can not yet used as wakeup
source. Most likely caused by missing irq_set_wake support in the
gpio/pinctrl driver.

Signed-off-by: Janne Grunau <j@jannau.net>

hoshinolina committed:
rust: drm: gem: shmem: Add DRM shmem helper abstraction

The DRM shmem helper includes common code useful for drivers which
allocate GEM objects as anonymous shmem. Add a Rust abstraction for
this. Drivers can choose the raw GEM implementation or the shmem layer,
depending on their needs.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/gem: Add a flag to control whether objects can be exported

Drivers may want to support driver-private objects, which cannot be
shared. This allows them to share a single lock and enables other
optimizations.

Add an `exportable` field to drm_gem_object, which blocks PRIME export
if set to false. It is initialized to true in
drm_gem_private_object_init.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: drm: gem: Add set_exportable() method

This allows drivers to control whether a given GEM object is allowed to
be exported via PRIME to other drivers.

hoshinolina committed:
rust: drm: mm: Add DRM MM Range Allocator abstraction

drm_mm provides a simple range allocator, useful for managing virtual
address ranges. Add a Rust abstraction to expose this module to Rust
drivers.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: dma_fence: Add DMA Fence abstraction

DMA fences are the internal synchronization primitive used for DMA
operations like GPU rendering, video en/decoding, etc. Add an
abstraction to allow Rust drivers to interact with this subsystem.

Note: This uses a raw spinlock living next to the fence, since we do
not interact with it other than for initialization.
TODO: Expose this to the user at some point with a safe abstraction.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: drm: syncobj: Add DRM Sync Object abstraction

DRM Sync Objects are a container for a DMA fence, and can be waited on
signaled, exported, and imported from userspace. Add a Rust abstraction
so Rust DRM drivers can support this functionality.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: drm: sched: Add GPU scheduler abstraction

The GPU scheduler manages scheduling GPU jobs and dependencies between
them. This Rust abstraction allows Rust DRM drivers to use this
functionality.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
*RFL import: kernel::user_ptr

Commit reference: 3dfc5ebff103

hoshinolina committed:
MISSING SIGNOFFS: rust: Add PAGE_SIZE constant to kernel crate

Computed based on the PAGE_SHIFT macro from C.

// Co-authored-by: Adam Bratschi-Kaye <ark.email@gmail.com>
// Co-authored-by: Miguel Ojeda <ojeda@kernel.org>
// Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
*RFL import: kernel::module_param

Commit reference: 3dfc5ebff103

hoshinolina committed:
*RFL import: kernel::io_mem

Commit reference: 3dfc5ebff103

hoshinolina committed:
*RFL import: kernel::delay

Commit reference: 3dfc5ebff103

hoshinolina committed:
*RFL import: macros::module params functionality & deps

Commit reference: 3dfc5ebff103

hoshinolina committed:
*RFL import: Rest of kernel::error::code::*

Commit reference: 3dfc5ebff103

hoshinolina committed:
rust: Allow feature allocator_api

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
*RFL import: kernel::KParamGuard & friends

Commit reference: 3dfc5ebff103

hoshinolina committed:
drm/asahi: Add the Asahi driver UAPI

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: bindings: Bind the Asahi DRM UAPI

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: io_pgtable: Add the Apple UAT format abstraction

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: macros: Add versions macro

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Add the Asahi driver for Apple AGX GPUs

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/scheduler: Clean up jobs when the scheduler is torn down.

drm_sched_fini() currently leaves any pending jobs dangling, which
causes segfaults and other badness when job completion fences are
signaled after the scheduler is torn down.

Explicitly detach all jobs from their completion callbacks and free
them. This makes it possible to write a sensible safe abstraction for
drm_sched, without having to externally duplicate the tracking of
in-flight jobs.

This shouldn't regress any existing drivers, since calling
drm_sched_fini() with any pending jobs is broken and this change should
be a no-op if there are no pending jobs.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/scheduler: Fix UAF in drm_sched_fence_get_timeline_name

A signaled scheduler fence can outlive its scheduler, since fences are
independencly reference counted. Therefore, we can't reference the
scheduler in the get_timeline_name() implementation.

Fixes oopses on `cat /sys/kernel/debug/dma_buf/bufinfo` when shared
dma-bufs reference fences from GPU schedulers that no longer exist.

Signed-off-by: Asahi Lina <lina@asahilina.net>

marcan committed:
HID: Bump maximum report size to 16384

This maximum is arbitrary. Recent Apple devices have some vendor-defined
reports with 16384 here which fail to parse without this, so let's bump
it to that.

This value is used as follows:

report->size += parser->global.report_size * parser->global.report_count;

[...]

/* Total size check: Allow for possible report index byte */
if (report->size > (max_buffer_size - 1) << 3) {
	hid_err(parser->device, "report is too long\n");
	return -1;
}

All of these fields are unsigned integers, and report_count is bounded
by HID_MAX_USAGES (12288). Therefore, as long as the respective maximums
do not overflow an unsigned integer (let's say a signed integer just in
case), we're safe. This holds for 16384.

Signed-off-by: Hector Martin <marcan@marcan.st>

hoshinolina committed:
iommu/io-pgtable: Hack in FW-RW/GPU-RO mode into UAT io_pgtable

This is ugly, we need a better way of expressing this.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
DO NOT MERGE: drm/asahi: Add an experimental UAPI extension

This lets us play around with unknown stuff from userspace, without
having to commit to adding it to the UAPI (or knowing whether it's safe
to expose at all).

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: render: Implement unknown value UAPI extension

hoshinolina committed:
rust: drm: gem: Allow pinning GEM object driver data

This requires type_alias_impl_trait.

Signed-off-by: Asahi Lina <lina@asahilina.net>

marcan committed:
HID: magicmouse: Handle touch controller resets on SPI devices

On at least some SPI devices (e.g. recent Apple Silicon machines), the
Broadcom touch controller is prone to crashing. When this happens, the
STM eventually notices and resets it. It then notifies the driver via
HID report 0x60, and the driver needs to re-enable MT mode to make
things work again.

This poses an additional issue: the hidinput core will close the
low-level transport while the device is closed, which can cause us to
miss a reset notification. To fix this, override the input open/close
callbacks and send the MT enable every time the HID device is opened,
instead of only once on probe. This should increase general robustness,
even if the reset mechanism doesn't work for some reason, so it's worth
doing it for USB devices too. MTP devices are exempt since they do not
require the MT enable at all.

Signed-off-by: Hector Martin <marcan@marcan.st>

hoshinolina committed:
rust: drm: device: Convert Device to AlwaysRefCounted

Switch from being a refcount wrapper itself to a transparent wrapper
around `bindings::drm_device`. The refcounted type then becomes
ARef<Device<T>>.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: hw,initdata: Initdata fixes for G14S

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: hw: Drop max_dies

This isn't available on t602x, so it was probably a bad guess. Just use
the known die count for the userspace info, from our HW info tables.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: initdata: Fix ver_info for G13 V13.5

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: fw/channels: Identify subpipe (?) field in tvb ack

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: fw/channels: Document more message types a bit

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: initdata: Fix 13.5 field position for G13X

Only affects unk_534 getting shifted over for G13X.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: initdata: Document more stuff

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: fw/microseq: Document bits of has_attachments

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: hw/t600x: Expand some mappings

Seems this changed in 13.5. Probably doesn't matter but should be safe
to always expand.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: hw/t8112: Make last IOMapping read-only

This matches 13.5 firmware and t60xx.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: render: Correct some values for G14

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: initdata: Initialize GpuStatsFrag properly

I don't know if this matters, but just in case...

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: slotalloc: Allow initializing empty slots

These slots can never be used.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: fw.channels: Add static assertions for message sizes

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: hw.t602x: Fixes for t6022

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: regs: Fix cluster count for G14D

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Enable probing for t6022

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Increase recursion limit

We're running into this again with the init macros...

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: initdata: Fixes for G14D

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: initdata,hw: Identify & set idle_off_standby_timer

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: initdata: Fudge t3 calculation a bit

This gives exact numbers for G14D on the test machine. It looks like
a rounding thing, but it's not clear exactly what the original logic
would be here... I'm guessing the macOS driver uses floats and a
different order of operations and that's why it comes up with a
different result sometimes.

Also fix an incorrectly named variable.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: gpu,hw: Fix array IOMappings

These arrays aren't consecutive, but rather are supposed to have a
stride and possibly also span multiple dies. This doesn't affect T8xxx,
but I have no idea how this ever worked for T60xx...

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: hw/t602x: Increase meta/preempt sizes for G14D

meta2,meta3 confirmed, meta4 conservative guess (it's small anyway). Not
sure yet if max_splits also needs to be doubled.

Compute preempt just conservatively doubled. Since that one is a bit
larger, use the smaller size for <= G14C.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Expose firmware version to userspace

Might be useful for something.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: alloc: Support tagging array allocs

It's hard to tell what a given array buffer is just from the type, so
add support for explicitly adding a u32 tag. This can help us
differentiate between allocs in the debug codepaths or when dumping
memory.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: alloc: Use tag as the guard marker

To more easily debug GPU/FW-side overreads, use the alloc tag to fill
the padding instead of using a constant.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: buffer: Add tags to all the buffers

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: initdata: Tag more arrays

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: queue/compute: Tag preempt buf

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: render: Tag AuxFB

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: workqueue: Tag GPU buf

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: buffer,render: Identify and provide layer meta buf

It looks like one of the "heapmeta" pointers is actually a layer
metadata pointer, that macOS just allocates contiguously with the
tilemap headers and heap meta buffers. Size seems to always be 0x100.

Let's allocate it after the heapmeta, which will make debugging easier.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: compute: Implement bindless samplers

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: render: Implement bindless samplers

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: fw,queue: Implement helper programs

Also expose no preemption flag (?) separately.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: render: Identify and set Z/S strides for layered rendering

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: queue: Quieten some debugs

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Add verbose UAPI error reporting

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: file: Remove sync limit

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: render: Remove sync TVB growth support

We decided we can't reasonably support this given fence forward progress
requirements.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: buffer: Complain more loudly about sync requests

This codepath is known broken and we have no plans to fix it. However,
it works most of the time, and it can work around bugs in the minimum
TVB size calculation formula. So if that ever happens, we still want it
to fix the problem, but complain loudly so we can fix the calculation.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Identify & add render VS spills flag

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Identify and allocate clustered layering metadata buf

Turns out multi-cluster machines also need a clustered buffer for
layered rendering.

Fixes layered rendering on G13X with barriers (I guess if you don't
flush memory this stays in some kind of cache and somehow doesn't
matter?).

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Fix Clippy complaints

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: render: Fix layered rendering on G14X

We were missing one register and one bit.

Signed-off-by: Asahi Lina <lina@asahilina.net>

marcan committed:
rust: Fix x86 build

Signed-off-by: Hector Martin <marcan@marcan.st>

hoshinolina committed:
drm/asahi: render, buffer: Fix layered rendering on G13X (again)

The meta1 buffer needs to be multiplied by the layer count. This does
not seem to be needed on G14X (where even the stride is not specified
anywhere and no other registers are affected by the layer count).

Signed-off-by: Asahi Lina <lina@asahilina.net>

marcan committed:
drm/asahi: initdata: New init chain API for rebase

Not going to split this up into fixups with all the reformatting...
Lina gets to do that on next rebase.

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
rust: Add missing impl_trait_in_assoc_type feature

Signed-off-by: Hector Martin <marcan@marcan.st>

hoshinolina committed:
rust: bindgen: Exclude list functions with unsupported ABIs

CONFIG_LIST_HARDENED causes certain list functions to use what bindgen
considers an unsupported ABI:

panicked at 'Invalid or unknown abi 14 for function
"__list_del_entry_valid_or_report" (Function { name:
"__list_del_entry_valid_or_report", mangled_name:
Some("__list_del_entry_valid_or_report"), link_name:
None, signature: TypeId(ItemId(38153)), kind: Function,
linkage: External })',
/usr/share/cargo/registry/bindgen-0.68.1/codegen/mod.rs:4195:17

Exclude these, since they are internal functions we don't use directly
anyway.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/shmem-helper: Add lockdep asserts to vmap/vunmap

Since commit 21aa27ddc582 ("drm/shmem-helper: Switch to reservation
lock"), the drm_gem_shmem_vmap and drm_gem_shmem_vunmap functions
require that the caller holds the DMA reservation lock for the object.
Add lockdep assertions to help validate this.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: helpers: Add bindings/wrappers for dma_resv

This is just for basic usage in the DRM shmem abstractions for implied
locking, not intended as a full DMA Reservation abstraction yet.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: workqueue: Work around ICE in Rust 1.74

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: queue: Alocate the NotifierList as shared

macOS does it this way, and the firmware needs it (no cache management!)

Fixes firmware crashes with piglit with high concurrency.

Signed-off-by: Asahi Lina <lina@asahilina.net>

marcan committed:
HID: magicmouse: Query device dimensions via HID report

For SPI/MTP trackpads, query the dimensions via HID report instead of
hardcoding values.

TODO: Does this work for the USB/BT devices? Maybe we can get rid of the
hardcoded sizes everywhere?

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
HID: transport: spi: Implement GET FEATURE

This is used for fetching trackpad dimensions.

Signed-off-by: Hector Martin <marcan@marcan.st>

hoshinolina committed:
drm/asahi: compute: Allow no preemption flag

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Identify and implement helper config register

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Check command structure sizes

Eventually this can be used to extend the structure at the end
backwards-compatibly, for cases where we missed core fields in the UAPI.
More discrete features should be implemented via extensions.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: microseq: Fix warnings

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Convert to GPUVM and implement more VM_BIND ops

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: drm: Add GPUVM Manager abstraction

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: init: Add default() utility function

Initializer for types with Default::default() implementations in init
context. This, by nature, only works for types which are not pinned.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: init: Disable clippy::undocumented_unsafe_blocks lint

The missing SAFETY comments should be fixed later...

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: io_buffer: Disable clippy::undocumented_unsafe_blocks lint

The missing SAFETY comments should be fixed later...

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: init: macros: Disable clippy::undocumented_unsafe_blocks lint

The missing SAFETY comments should be fixed later...

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: module_param: Disable clippy::undocumented_unsafe_blocks lint

The missing SAFETY comments should be fixed later...

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: print: Disable clippy::undocumented_unsafe_blocks lint

The missing SAFETY comments should be fixed later...

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
kernel: str: Add/fix SAFETY comments

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: user_ptr: Disable clippy::undocumented_unsafe_blocks lint

The missing SAFETY comments should be fixed later...

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: kernel: workqueue: Disable clippy::undocumented_unsafe_blocks lint

The missing SAFETY comments should be fixed later...

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: Enable clippy::undocumented_unsafe_blocks lint

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/gpuvm: Add drm_gpuvm_bo_unmap()

Analogous to drm_gpuvm_bo_unmap_ops_create, this is a callback-driven
unmap function for a given BO.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Don't lock up when unmapping PTEs fails

If a bug causes PTEs to be unmapped twice, the unmap loop gets stuck
spamming WARNs forever. Just skip a page and try again so we can make
forward progress.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Refactor address types

VAs are u64, PAs and sizes are usize.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: util: Add RangeExt helpers for Range<T>

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: mmu: Convert to using Range

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Move the unknown dummy page to the top of the address space

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Convert more ranges to Range<>

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: file: Update to newer VM_BIND API

Signed-off-by: Asahi Lina <lina@asahilina.net>

jannau committed:
Revert "rust: remove `params` from `module` macro example"

This reverts commit 19843452dca40e28d6d3f4793d998b681d505c7f.

Signed-off-by: Janne Grunau <j@jannau.net>

hoshinolina committed:
drm/asahi: alloc: Do not allocate memory to free memory

The existing garbage mechanism could allocate a relatively unbounded vec
when freeing garbage, which was hurting memory exhaustion scenarios.
The only reason we need that buffer is to move garbage out of the lock
so we can drop it without deadlocks. Replace it with a 128-size
pre-allocated garbage buffer, and loop around reusing it.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: mmu: Fix lockdep issues with GpuVm

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: drm: gem: shmem: Add share_dma_resv() function

Allow a GEM object to share another object's DMA reservation, for use
with drm_gpuvm. To keep memory safety, we hold a reference to the GEM
object owning the resv, and drop it when the child object is freed.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Implement GEM objects sharing a single DMA resv

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Signal soft fault support to userspace

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: queue: Split into Queue and QueueInner

Work around mutability issues when entity.new_job() takes a mutable
reference to the entity by moving all the fields used by the
submit_render() and submit_compute() functions to an inner struct,
eliminating the double-mutable-borrow.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: workqueue: Tweak some debug logs

Silence some logs that are possible and harmless in rare cases, add more
debug to another log that is questionable.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Enable soft faults by default

Signed-off-by: Asahi Lina <lina@asahilina.net>

jannau committed:
drm/asahi: run rustfmt

Signed-off-by: Janne Grunau <j@jannau.net>

hoshinolina committed:
drm/asahi: Fix u32 mult overflow on large tilebufs/TPCs

Signed-off-by: Asahi Lina <lina@asahilina.net>

jannau committed:
rust: block: rnull: Adapt to asahi's rust base branch

Signed-off-by: Janne Grunau <j@jannau.net>

hoshinolina committed:
drm/asahi: Workqueue: Add more debug

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Fix event tracking when JobSubmission is dropped

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: gpu: Show unknown field in timeouts

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
FIXUP: drm/asahi: firmware fixes

was "fixup! drm/asahi: Add the Asahi driver for Apple AGX GPUs" but
doesn't apply cleanly.

hoshinolina committed:
fixup! drm/asahi: Add the Asahi driver UAPI

hoshinolina committed:
drm/asahi: Handle channel errors

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: event: Initialize stamps to different values

Makes debugging a bit easier.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: workqueue,queue: More debug

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: workqueue: Fix "Cannot submit, but queue is empty?" bug

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Add robust_isolation kernel parameter

This only allows binding one VM context at once, which serializes GPU
usage between VMs and therefore prevents one faulting VM from affecting
others.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Clean up jobs in a workqueue

This eliminates a potential deadlock under load and improves the fence
signaling situation (for when we have a shrinker).

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
FIXUP: drm/asahi: firmware fixes 2

was "fixup! drm/asahi: Add the Asahi driver for Apple AGX GPUs" but
doesn't apply cleanly.

hoshinolina committed:
drm/asahi: HACK: Disable compute preemption for now

Possibly because we don't have support in the helper program, this is
broken and causes channel errors. Hack in high priority for now, which
works around it.

Use debug_flags 0x1000000000000 to re-enable for testing.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Align kernel range to buffer::PAGE_SIZE

We only require alignment to the UAT page size from userspace, but
internally we need more, so just align it if userspace gives us lower
alignment.

Signed-off-by: Asahi Lina <lina@asahilina.net>

jannau committed:
HID: transport: spi: apple: Increase receive buffer size

The SPI receive buffer is passed directly to hid_input_report() if it
contains a complete report. It is then passed to hid_report_raw_event()
which computes the expected report size and memsets the "missing
trailing data up to HID_MAX_BUFFER_SIZE (16K) or
hid_ll_driver.max_buffer_size (if set) to zero.

Signed-off-by: Janne Grunau <j@jannau.net>

hoshinolina committed:
drm/asahi: Implement missing ASAHI_BIND_OP_UNBIND

Trivial now that we have GPUVM.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: Add time_namespace helpers

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Implement ASAHI_GET_TIME

Signed-off-by: Asahi Lina <lina@asahilina.net>

WhatAmISupposedToPutHere committed:
dmaengine: apple-admac: Avoid accessing registers in probe

The ADMAC attached to the AOP has complex power sequencing, and is
power gated when the probe callback runs. Move the register reads
to other functions, where we can guarantee that the hardware is
switched on.

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

WhatAmISupposedToPutHere committed:
soc: apple: rtkit: Add apple_rtkit_has_endpoint()

To be used by RTKit consumers to check if an endpoint
is present and should be enabled.

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

WhatAmISupposedToPutHere committed:
soc: apple: rtkit: Add tracekit endpoint.

This system endpoint is advertised by AOP and also needs to be
turned on for it to function.

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

WhatAmISupposedToPutHere committed:
fixup! rust: soc: apple: rtkit: Add Apple RTKit abstraction

WhatAmISupposedToPutHere committed:
rust: alloc: Add .resize() to Vec

Similar in function to the one from std, but takes GFP flags.

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

WhatAmISupposedToPutHere committed:
rust: Add helper for memcpy_toio

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

metaspace committed:
rust: add dma pool and coherent allocator

Based on https://github.com/wedsonaf/linux/commit/02541e65a7e778c0049fed86ae49302bc07abed3

WhatAmISupposedToPutHere committed:
rust: WIP(?): Various bits of rust glue for AOP series

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

WhatAmISupposedToPutHere committed:
soc: apple: Add support for the AOP co-processor.

This is the base device for a multi-function co-processor present
on certain Apple SoCs. On M-series Macs it is in charge of internal
microphones, and various environmental sensors.

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

WhatAmISupposedToPutHere committed:
rust: alloc: kbox: Add AsRef implementation to Box

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

WhatAmISupposedToPutHere committed:
iio: common: Add AOP sensor drivers.

The AOP co-processor present on certain Apple SoCs exposes
various environmental sensors as "HID" (really not) devices.
Add drivers for the ambient light and lid angle sensors
exposed that way.

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

WhatAmISupposedToPutHere committed:
arm64: dts: apple: Add AOP and subdevices.

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

jannau committed:
fixup! drm/asahi: Add the Asahi driver for Apple AGX GPUs

jannau committed:
fixup! *RFL import: kernel::user_ptr

jannau committed:
fixup! rust: dma_fence: Add DMA Fence abstraction

jannau committed:
fixup! *RFL import: kernel::module_param

jannau committed:
fixup! rust: drm: sched: Add GPU scheduler abstraction

jannau committed:
fixup! rust: of: Add OF node abstraction

jannau committed:
fixup! rust: drm: Add GPUVM Manager abstraction

jannau committed:
fixup! drm/asahi: Convert to GPUVM and implement more VM_BIND ops

jannau committed:
rust: add unstable feature associated_type_defaults

Previously imported through "*RFL import: kernel::platform".

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
fixup! rust: drm: mm: Add DRM MM Range Allocator abstraction

jannau committed:
fixup! rust: platform: add `ioremap_resource` and `get_resource` methods

jannau committed:
fixup! rust: kernel: platform: Add Device.set_dma_masks()

jannau committed:
fixup! rust: drm: gem: shmem: Add DRM shmem helper abstraction

jannau committed:
fixup! rust: io_pgtable: Add the Apple UAT format abstraction

Required after commit 47f218d10895 ("iommu/amd: Store the nid in
io_pgtable_cfg instead of the domain").

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
fixup! drm/asahi: Expose firmware version to userspace

jannau committed:
fixup! drm/asahi: initdata: New init chain API for rebase

jannau committed:
fixup! drm/asahi: alloc: Do not allocate memory to free memory

jannau committed:
fixup! drm/asahi: Clean up jobs in a workqueue

jannau committed:
fixup! drm/asahi: Convert more ranges to Range<>

jannau committed:
fixup! drm/asahi: run rustfmt

jannau committed:
rust: of: Add parse_phandle_by_name()

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
HACK: rust: of: add as_raw()

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
rust: platform: Add platform::Device::get_device()

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
HACK: rust: drm: device: Add device initialization without data

Required by the asahi driver.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
HACK: rust: drm: device: Add pointless device() method

To be dropped when use in asahi driver is removed.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/asahi: Modify debug dev printk macros to avoid adding ".as_ref()"

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/asahi: gpu: Force Box move with manual Box<T>::into_inner()

TODO: Investigate why this doesn't work automatically.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/asahi: gpu: Get resources directly through DRM pricate data

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/asahi: mmu: Move closer to rust of integration

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/asahi: regs: Get device ref from platform dev

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/asahi: driver: Port to new device/driver registration

As proposed in
https://lore.kernel.org/rust-for-linux/20241022213221.2383-1-dakr@kernel.org/

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
rust: macros: module: Functionally revert commit 7044dcff8301

The reverts commit 7044dcff8301 ("rust: macros: fix soundness issue in
`module!` macro"). Restores old "*RFL import: macros::module params
functionality & deps" functionality.

Fixes: 7044dcff8301 ("rust: macros: fix soundness issue in `module!` macro")
Signed-off-by: Janne Grunau <j@jannau.net>

WhatAmISupposedToPutHere committed:
rust: soc: apple: Add Apple mailbox abstractions

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

WhatAmISupposedToPutHere committed:
rust: WIP(?): Various bits of rust glue for sep

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

WhatAmISupposedToPutHere committed:
soc: apple: Add SEP driver.

This is a co-processor in charge of various security-related
features on Apple SoCs. This driver only boots the firmware,
which is needed to unlock the mic secure disable on certain
laptop models.

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

WhatAmISupposedToPutHere committed:
arm64: dts: apple: Add SEP device tree nodes.

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

jacksongoode commented on arch/arm64/boot/dts/apple/t602x-die0.dtsi:
```diff
@@ -585,7 +700,7 @@
 
 	admac: dma-controller@39b400000 {
-		compatible = "apple,t6020-admac", "apple,admac";
+		compatible = "apple,t6020-admac2", "apple,t6020-admac", "apple,admac", "apple,admac";
```
One of these should be `apple,admac2`, right?

jannau committed:
rust: drm: sched: Use &mut for ArmedJob.fences()

Fixes build error after Box -> KBox conversion.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
fixup! rust: drm: device: Convert Device to AlwaysRefCounted

jannau committed:
fixup! drm/asahi: alloc: Use tag as the guard marker

jannau committed:
fixup! drm/asahi: Handle channel errors

jannau committed:
fixup! drm/asahi: hw: Drop max_dies

jannau committed:
fixup! drm/asahi: gpu: Show unknown field in timeouts

jannau committed:
fixup! drm/asahi: file: Update to newer VM_BIND API

jannau committed:
fixup! drm/asahi: Don't lock up when unmapping PTEs fails

jannau committed:
fixup! drm/asahi: workqueue,queue: More debug

WhatAmISupposedToPutHere committed:
rust: device: Add Device::parent()

Just a wrapper over dev->parent.

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

WhatAmISupposedToPutHere committed:
rust: Only export module id tables if building as module

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

WhatAmISupposedToPutHere committed:
ASoC: apple: Add aop_audio driver.

Apple SoCs have their microphones connected to the AOP
co-processor, in order to among other things implement
the "voicetrigger" functionality. Add a driver for the
"High power audio input" AOP endpoint.

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>

hoshinolina committed:
drm/asahi: gpu: Collect garbage for private/gpuro together

Avoids double firmware flushes

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: alloc: Be more verbose about failures

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
rust: alloc: Fix `ArrayLayout` allocations

We were accidentally allocating the *square* of the array length due to
a variable shadowing mishap.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: gpu: Add a max object count garbage limit

This ensures the garbage Vec does not grow beyond what is reasonable,
and probably reduces jank by doing more smaller GCs instead of big ones.

Signed-off-by: Asahi Lina <lina@asahilina.net>

jannau committed:
Merge branch 'bits/010-soc' into asahi-wip

jannau committed:
Merge branch 'bits/030-misc' into asahi-wip

jannau committed:
Merge branch 'bits/000-devicetree' into asahi-wip

jannau committed:
Merge branch 'bits/020-dart' into asahi-wip

jannau committed:
Merge branch 'bits/005-maintainers' into asahi-wip

jannau committed:
Merge branch 'bits/050-nvme' into asahi-wip

jannau committed:
Merge branch 'bits/070-audio' into asahi-wip

jannau committed:
Merge branch 'bits/080-wifi' into asahi-wip

jannau committed:
Merge branch 'bits/060-spi' into asahi-wip

jannau committed:
Merge branch 'bits/040-dwc3' into asahi-wip

jannau committed:
Merge branch 'bits/130-cpufreq' into asahi-wip

jannau committed:
Merge branch 'bits/090-spi-hid' into asahi-wip

jannau committed:
Merge branch 'bits/110-smc' into asahi-wip

jannau committed:
Merge branch 'bits/120-spmi' into asahi-wip

jannau committed:
Merge branch 'bits/150-xhci-firmware' into asahi-wip

jannau committed:
Merge branch 'bits/140-pci' into asahi-wip

jannau committed:
Merge branch 'bits/190-rust' into asahi-wip

jannau committed:
Merge branch 'bits/170-atcphy' into asahi-wip

jannau committed:
Merge branch 'bits/180-sio' into asahi-wip

jannau committed:
Merge branch 'bits/200-dcp' into asahi-wip

jannau committed:
Merge branch 'bits/230-touchbar' into asahi-wip

jannau committed:
Merge branch 'bits/220-tso' into asahi-wip

jannau committed:
Merge branch 'bits/240-isp' into asahi-wip

jannau committed:
Merge branch 'bits/210-gpu' into asahi-wip

chadmed commented on sound/soc/apple/aop_audio.rs:
```diff
@@ -0,0 +1,668 @@
+            copy_str(&mut target[name.len()..], chassis.as_ref());
+        }
+        unsafe {
+            chassis[0].make_ascii_uppercase();
```
Is this really necessary if the device tree is correct? We control the DTs, we should just make sure the value in the DT is correct.

chadmed commented on sound/soc/apple/aop_audio.rs:
```diff
@@ -0,0 +1,668 @@
+            let target = (*this.0).driver.as_mut();
+            copy_str(target, name.as_ref());
+            copy_str(&mut target[name.len()..], chassis.as_ref());
+        }
```
Is doing this twice intentional? Also, given that we set `snd_card.driver` to something entirely different below, do either of these blocks actually do anything?

chadmed commented on sound/soc/apple/aop_audio.rs:
```diff
@@ -0,0 +1,668 @@
+            let target = (*this.0).driver.as_mut();
+            copy_str(target, chassis.as_ref());
+            chassis.truncate(chassis.len() - name.len());
+        }
```
This should just be `aop` anyway.

chadmed reviewed (changes requested):
Just a couple of nits in the driver itself, but apart from those we need to make sure we're setting:
- `snd_card.shortname` (e.g. `MacBook Pro J314 HPAI`)
- `snd_card.longname` (e.g. `MacBook Pro J314 High-Power Audio Interface`)
- `snd_card.id` (e.g. `AppleJ314HPAI`)

This will align with what we do for macaudio and DCP, and makes doing UCM and DSP rules easier.

jannau commented:
I've merged everything except the device tree changes into bits/ branches. Most of it into `bits/190-rust` and `bits/250-aop` but fixups in there respective bits/* branches. I'm not happy with the split-up as merging `bits/250-aop` needs manual conflict resolution. I can't think of anything better though. Will get easier when we upstream more changes.

I did some minimal fixups. Can you please send 85e9396f80328f839c8864308af9ba91d128d970 ("dmaengine: apple-admac: Avoid accessing registers in probe") (without the compatible addition) as soon as possible. If I remember correctly special power sequencing requirements do not warrant a different  compatible. If we fix upstream now we can upstream aop-admac with the same compatible string later. Downstream we will keep "apple,admac2" for now.

For commit b63a09090638 ("rust: Only export module id tables if building as module") please reply to https://lore.kernel.org/rust-for-linux/20241022213221.2383-5-dakr@kernel.org/

clintaire commented:
Can the necessary configurations for `snd_card` fields be abstracted into a reusable module or script to avoid repetitive work and standardize the approach across different hardware?

WhatAmISupposedToPutHere commented:
> Can the necessary configurations for `snd_card` fields be abstracted into a reusable module or script to avoid repetitive work and standardize the approach across different hardware?

Sorry i am not quite sure i understand. Are you asking as to whether there will be rust bindings for alsa? If so - yes, eventually, and those may or may not come from me.

rosvall commented:
> Draft, since the mic output is basically useless without correct userspace configs (coming soontm).

I've played a with this for a bit, and at least on my m1 macbook air, the sound is absolutely usable!

Here's a short 3 channel recording: [mic_test_j313.wav.zip](https://github.com/user-attachments/files/18252942/mic_test_j313.wav.zip)

Please merge the devicetree bindings and let people have some fun with it.

WhatAmISupposedToPutHere commented:
> > Draft, since the mic output is basically useless without correct userspace configs (coming soontm).
> 
> I've played a with this for a bit, and at least on my m1 macbook air, the sound is absolutely usable!
> 
> Here's a short 3 channel recording: [mic_test_j313.wav.zip](https://github.com/user-attachments/files/18252942/mic_test_j313.wav.zip)
> 
> Please merge the devicetree bindings and let people have some fun with it.

How much did you boost the signal?

rosvall commented:
> > > Draft, since the mic output is basically useless without correct userspace configs (coming soontm).
> > 
> > 
> > I've played a with this for a bit, and at least on my m1 macbook air, the sound is absolutely usable!
> > Here's a short 3 channel recording: [mic_test_j313.wav.zip](https://github.com/user-attachments/files/18252942/mic_test_j313.wav.zip)
> > Please merge the devicetree bindings and let people have some fun with it.
> 
> How much did you boost the signal?

It's not boosted. I've just set configuration profile to "Pro Audio" and gain to 0dB in pavucontrol, and recorded with `arecord -D pipewire -d 10 -r 48000 -v -f FLOAT_LE -c 3 mic_test_j313.wav`

WhatAmISupposedToPutHere commented:
> > > > Draft, since the mic output is basically useless without correct userspace configs (coming soontm).
> > > 
> > > 
> > > I've played a with this for a bit, and at least on my m1 macbook air, the sound is absolutely usable!
> > > Here's a short 3 channel recording: [mic_test_j313.wav.zip](https://github.com/user-attachments/files/18252942/mic_test_j313.wav.zip)
> > > Please merge the devicetree bindings and let people have some fun with it.
> > 
> > 
> > How much did you boost the signal?
> 
> It's not boosted. I've just set configuration profile to "Pro Audio" and gain to 0dB in pavucontrol, and recorded with `arecord -D pipewire -d 10 -r 48000 -v -f FLOAT_LE -c 3 mic_test_j313.wav`

Well, this is not how it works on, say j314, where a boost of 50-ish dba is required for the signal to be audible.

rosvall commented:
> Well, this is not how it works on, say j314, where a boost of 50-ish dba is required for the signal to be audible.

Is it broken on every j314?

WhatAmISupposedToPutHere commented:
> > Well, this is not how it works on, say j314, where a boost of 50-ish dba is required for the signal to be audible.
> 
> Is it broken on every j314?

At least two of them. j415 also needs 20-30 dba of boost.

chadmed commented:
That mic signal is useless and almost silent. J313 will need boosting too.

jannau commented:
device nodes (disabled) included in asahi-6.12.8-1, going to close this PR
trivial patches to enable the mics are in https://github.com/jannau/linux/commits/mics-asahi-6.12.8-1/

changes compared to this PR:
- enable mics on M1 imacs
- enable sep only on t8103-j*
